File: src/main/java/net/jgp/books/spark/ch14/lab910_addition/AdditionApp.java
package net.jgp.books.spark.ch14.lab910_addition;

import static org.apache.spark.sql.functions.callUDF;
import static org.apache.spark.sql.functions.col;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

/**
 * Additions via UDF.
 * 
 * @author jgp
 */
public class AdditionApp {

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    AdditionApp app = new AdditionApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Addition")
        .master("local[*]")
        .getOrCreate();
    spark.udf().register(
        "add_int",
        new IntegerAdditionUdf(),
        DataTypes.IntegerType);
    spark.udf().register(
        "add_string",
        new StringAdditionUdf(),
        DataTypes.StringType);

    Dataset<Row> df = createDataframe(spark);
    df.show(false);

    df = df
        .withColumn(
            "concat",
            callUDF("add_string", col("fname"), col("lname")));
    df.show(false);
    
    df = df
        .withColumn(
            "score",
            callUDF("add_int", col("score1"), col("score2")));
    df.show(false);
  }

  private static Dataset<Row> createDataframe(SparkSession spark) {
    StructType schema = DataTypes.createStructType(new StructField[] {
        DataTypes.createStructField(
            "fname",
            DataTypes.StringType,
            false),
        DataTypes.createStructField(
            "lname",
            DataTypes.StringType,
            false),
        DataTypes.createStructField(
            "score1",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "score2",
            DataTypes.IntegerType,
            false) });

    List<Row> rows = new ArrayList<>();
    rows.add(RowFactory.create("Jean-Georges", "Perrin", 123, 456));
    rows.add(RowFactory.create("Jacek", "Laskowski", 147, 758));
    rows.add(RowFactory.create("Holden", "Karau", 258, 369));

    return spark.createDataFrame(rows, schema);
  }
}

File: src/main/python/lab910_addition/additionApp.py
"""
 Additions via UDF.

 @author rambabu.posa
"""
from pyspark.sql import (SparkSession, functions as F)
from pyspark.sql.types import (StructType,StructField,
                               StringType, IntegerType, BooleanType)

def createDataframe(spark: SparkSession):
    schema = StructType([
        StructField('fname', StringType(), False),
        StructField('lname', StringType(), False),
        StructField('score1', IntegerType(), False),
        StructField('score2', IntegerType(), False)
    ])

    rows = [
        ("Jean-Georges", "Perrin", 123, 456),
        ("Jacek", "Laskowski", 147, 758),
        ("Holden", "Karau", 258, 369)
    ]
    return spark.createDataFrame(rows, schema)


def main(spark):
    df = createDataframe(spark)
    df.show(truncate=False)

    add_string = F.udf(lambda a,b: a+b, StringType())
    add_int    = F.udf(lambda a,b: a+b, IntegerType())

    df = df.withColumn("concat",
            add_string(F.col("fname"), F.col("lname")))
    df.show(truncate=False)

    df = df.withColumn("score",
               add_int(F.col("score1"), F.col("score2")))
    df.show(truncate=False)

if __name__ == '__main__':
    # Creates a session on a local master
    spark = SparkSession.builder.appName("Addition") \
        .master("local[*]").getOrCreate()
    # Comment this line to see full log
    spark.sparkContext.setLogLevel('error')
    main(spark)
    spark.stop()